### PR TITLE
Add loop-ownership check to ValueTraceLoopComplete

### PR DESCRIPTION
**Pull request by haroldulrich**
# Description

## ❓ What does this PR do?

Adds the same loop-ownership check that `ValueTraceLoopIncomplete` already has to `ValueTraceLoopComplete`, making enclosing loops correctly see `tshape_unknown` instead of the pre-loop type. The optimizer no longer treats certain variables as None after nested for-loops.

Previously, when an outer loop asked a `ValueTraceLoopComplete` from an inner for loop about its shape alternatives, there was no `emitShapeAlternativesForLoop` override — so it fell through to a default that returned the pre-loop shape (None). The optimizer then concluded the variable was still None after the nested loops and mis-optimized all resulting operations. The fix makes the loop-complete trace distinguish "my loop" from "someone else's loop" when reporting shapes.

## 🧐 Why was it initiated? Any relevant Issues?

-  #3696

## 🧱 Type of Change

- [x] 🐛 Bug fix (non-breaking change which fixes an issue)
- [x] ✨ New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as
  expected)
- [ ] 🧹 Refactoring (no functional changes, no api changes)
- [ ] 🏗️ Build / CI System
- [ ] 📚 Documentation Update

## ✅ PR Checklist

- [x] **Correct base branch selected**: Should be `develop` branch.
- [x] **Formatting**: Enabled commit hook or executed `./bin/autoformat-nuitka-source`.
- [ ] **Tests**: All tests still pass. (See
  [Running the Tests](https://nuitka.net/doc/developer-manual.html#running-the-tests)).
  - CI actions cover the basics, but manual verification is encouraged.
- [x] **New Coverage**: New features or fixed regressions are covered via new tests.
- [ ] **Documentation**: Documentation updates are included for new or changed features.

## 🤖 AI Generated Code Policy

- [x] **Detection**: This PR contains AI generated code.
  - [x] **Issue First**: I have created an issue explaining the problem *before* using AI to
    generate a fix, ensuring the direction is correct.
  - [x] **Documentation**: I have provided the prompts used to generate the code (non-optional).
  - [x] **Verification**: I have **manually verified** the AI generated code.
    > **Note**: AI generated code is welcome, but it must be peer-reviewed and understood by the
    > submitter. Blindly copying AI output without understanding is not acceptable.
  - [x] **Evidence**: I have included test evidence that the change is effective.

## Prompts Used To Generate Code

- Analyze Nuitka issue #3696 and figure out why nested for-loops cause the optimizer to lose their type shape data. A fix already exists with how ValueTraceLoopIncomplete overrides emitShapeAlternativesForLoop with a loop-ownership check. But the bug can still be reproduced with the author's example in the issue report. Figure out the root cause and if other ValueTrace code is missing a similar approach, or if the issue is deeper and more widespread. When you reach near 100% confidence, make a targeted fix to the affected areas.
- Write a test module in tests/basics/ that reproduces the original author's issue. Cover as many variations and edge cases as appropriate, such as nested for-while, for-for, etc. loops. Wrap each test method in a try/except so they all run at least once, add a pass/fail counter, and print a short summary with the pass/fail results at the end.
- Tests that pass before and after the fix may not provide useful information. Trim down these excess bloat tests that pass before and after the fix. Conversely, if a test passes before and after the fixes, it may provide useful insights and can serve as a canary for future development. Only keep tests that pass before and after the fixes if they contain genuinely useful analysis. Follow Nuitka's test convention where CPython and Nuitka stdout must match exactly.
- Now write a similar test module for operations performed on variables after nested for-loops complete, like list concatenation, arithmetic, string operations, etc. Follow the same pattern: individual try/except per test so they each run at least once, pass/fail counter, and a summary at the end. Then trim to only tests that fail without the fix plus a lean set of useful canaries.

#### Comparison of test runs (NestedLoopInferenceTest.py, NestedLoopPostOpsTest.py) on each branch
- Running new tests on `emit-shape-alternatives`
  - NestedLoopInferenceTest.py
    Ran 1 test cases (0 failures).
  - NestedLoopPostOpsTest.py
    Ran 1 test cases (0 failures).
- Running new tests on `develop`
  - NestedLoopInferenceTest.py
    +Summary: 6 passed, 4 failed, 10 total
  - NestedLoopPostOpsTest.py
    +Summary: 7 passed, 13 failed, 20 total

`ValueTraceLoopIncomplete` already had the correct override. It checks `self.loop_node` is `loop_node` and falls back to `tshape_unknown` for enclosing loops. But `ValueTraceLoopComplete` — the trace that represents a finished loop analysis was missing the same override. So it fell through to the base class default in `ValueTraceBase`, which just calls `self.getTypeShape().emitAlternatives(emit)` without checking which loop is asking.

______________________________________________________________________

## Summary by Sourcery

Fix type shape propagation across nested loops and add regression coverage for nested-loop inference and post-loop operations.

Bug Fixes:
- Ensure ValueTraceLoopComplete reports tshape_unknown to enclosing loops instead of reusing the pre-loop type, preventing misoptimization of variables after nested loops.

Enhancements:
- Delegate loop shape emission through intermediate value traces so enclosing loops see the correct shape information.

Tests:
- Add NestedLoopInferenceTest to exercise type inference for variables reassigned from None inside various nested loop patterns and containers.
- Add NestedLoopPostOpsTest to cover operations performed on variables after nested loops, including arithmetic, concatenation, unpacking, comprehensions, and other builtin uses.

### CODE REVIEW

**Comment by kayhayen** [comment]
This one is excellent in complicance with my PR template. Thanks for the effort, this is extremely helpful.

When reviewing the change, I think I spotted that the overload that delegates to "previous" cannot be right, and actually would have to be totally wrong. But the point about not doing the same check for complete stands absolutely tall.

I will check out your branch and see if I can get rid of the wrong part. In typical AI fashion, it may have done some symptom overriding with that delegation, and when that didn't help, find the actual problem, fix it, and leave the other code in. But I might also be very confused, and indeed it will reveal that there are more issues.

For me to use the tests or do a merge, it will require a CLA to be signed. Let me know if you are willing to do that.

File: tests/basics/NestedLoopInferenceTest.py [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+    out = []
+    if collected is not None:
+        for x in collected:
+            out.append(x)
+    print("for+for list iterate:", out)
+
+
+run_test(test_for_for_list_iterate)
+
+
```
[the remark this answers is not in the document]
Comment (kayhayen) [reply tests/basics/NestedLoopInferenceTest.py:48]: Seems like a good idea for increased coverage.

**Comment by kayhayen** [comment]
So, the overload is on the escape trace. I am a bit surprised, that would even make a difference, but it does. The reason is probably that for generic traces, we ignore the loop, if it's one that takes previous seriously. Unknown traces, assignment, traces, all but escape traces, do not, which is why it's also only needed there.

But the complete loop has complete information, and can be applied for all loops and doesn't have the overload on purpose, and we can remove that without failing tests. In fact, I believe it would only serve to degrade optimization of nested loops.

**Comment by kayhayen** [comment]
I pushed the correction, since I would live to include this into 4.0.1 @hygtfrde let me know about your stance on signing the CLA.

**Comment by kayhayen** [comment]
Ah, I am seeing the CLA bot has trouble identifying the GitHub account for the commits, probably therefore has not prompted you to do it.

**Comment by haroldulrich** [comment]
CLA was agreed, I see "You have agreed to the CLA for Nuitka/Nuitka" on the link https://cla-assistant.io/Nuitka/Nuitka?pullRequest=3757

**Comment by haroldulrich** [comment]
On a side note, I was looking at other places where loop-ownership checks could be missing, and how similar patterns could emerge in other areas. After prompting Claude to scour the codebase for related structures where this issue could manifest, It provided this response:
"ValueTraceMerge (line 826) — handles merge points where control flow converges (e.g., after if/else). It doesn't delegate to self.previous for getTypeShape(), so it doesn't have the same bug. But it does iterate through multiple previous traces and compare shapes. If a merge point sits between nested loops, the base default would emit the merged shape without loop-awareness. Right now this doesn't seem to cause issues, but it's the most likely place for a future manifestation of a similar bug — a variable that goes through an if/else branch inside a nested loop might get its shape merged incorrectly at the merge point when an outer loop asks about it."
I'm looking into what other compilers did for situations like these.

**Comment by kayhayen** [comment]
I have to agree, the base method of the method, cannot just drop the argument with no consideration of the concret node, that is just asking for problems, and I have something in mind for the merge trace. It will have to ask its branches to emit for the loop, and then try to compactify the result. For now, it should be safe, since "is" identity is not going to happen for the loop shapes at all.

If you find prior art of dynamic language loop analysis for types, I would be very interested. I did not find any prior art when I had attempted to search it, but AI might be better at finding stuff that my Google-fu at the time.

I am adding a TODO about it now. The type shapes of merge traces will get a re-work at some point, and I like to annotate my bugs with TODO items.

**Comment by kayhayen** [comment]
I think the bot doesn't connect the github account with the email address, and therefore doesn't consider the commits of that email address to be covered.

**Comment by kayhayen** [comment]
I see it changed its mind, let me click merge there, now, I will however push it away with changes from the hotfix and add the tests for 4.1 only.
